src/main/java/com/hivemq/mqtt/message/pool/FreePacketIdRanges.java

```diff
@@ -0,0 +1,187 @@
+ * integer ids that are NOT currently assigned to any object using this instance of {@link FreePacketIdRanges}.
+ * The lower end of the {@link Range} instance is included whereas the upper end is excluded from the interval.
+ * That is, the {@link Range} instance with start 42 and end 42 contains only one ID - 42.
+ * The {@link Range} instance with start 10 and end 12 contains only two IDs: 10 and 11.
+ * <p>
+ * Initially, there is only one contiguous {@link Range} of IDs. The IDs are assigned starting from the lowest one in
+ * that interval.
+ * Upon assignment, the interval's lower end is incremented (interval size reduces from below).
+ * When the ID is returned, it either joins one of the existing {@link Range} intervals in the list (if it is adjacent)
+ * or it forms a new {@link Range} that is added to the list.
+ * <p>
+ * This class is NOT thread-safe.
+ * <a
+ * href="https://github.com/hivemq/hivemq-mqtt-client/blob/master/src/main/java/com/hivemq/client/internal/util/Ranges.java">The
+ * original implementation in the HiveMQ Java Client.</a>
+ */
+public class FreePacketIdRanges {
+
+    private static final int MIN_ALLOWED_MQTT_PACKET_ID = 1;
+
+    @VisibleForTesting
+    public static final int MAX_ALLOWED_MQTT_PACKET_ID = 65_535;
+
+    private @NotNull Range rootRange;
+
+    public FreePacketIdRanges() {
+        rootRange = new Range(MIN_ALLOWED_MQTT_PACKET_ID, MAX_ALLOWED_MQTT_PACKET_ID + 1);
+    }
+
+    /**
+     * Provides a new ID that is not currently allocated.
+     *
+     * @return a new ID if available in any of the ranges or {@link NoMessageIdAvailableException} if ran out of IDs.
+     */
+    public int takeNextId() throws NoMessageIdAvailableException {
+        if (rootRange.start == rootRange.end) {
+            throw new NoMessageIdAvailableException();
+        }
+
+        final int id = rootRange.start;
+        rootRange.start++;
+        if ((rootRange.start == rootRange.end) && (rootRange.next != null)) {
+            rootRange = rootRange.next;
+        }
+        return id;
+    }
+
+    /**
+     * Provides the requested ID if it is available or some other ID otherwise.
+     *
+     * @param id an ID that the caller attempts to take.
+     * @return the requested {@param id} if it is available in one of the {@link Range} intervals or otherwise some
+     *         other free ID.
+     */
+    public void takeIfAvailable(final int id) {
+        Preconditions.checkArgument(id >= MIN_ALLOWED_MQTT_PACKET_ID && id <= MAX_ALLOWED_MQTT_PACKET_ID,
+                "Attempting to take an ID %s that is outside the valid packet IDs range.",
+                id);
+
+        Range current = rootRange;
+        Range prev = null;
+
+        while (current != null) {
+            if (id < current.start) {
+                return; // since the ranges are traversed in increasing order of IDs, the given id will not be found
```

Comment:
what happens is this case? is the caller allowed to take the id if it's not there?

Reply:
ok, I see, this is to prevent using a ID used by the client

Reply:
Specifically this exit condition is just to exit the method without requiring to look at every range; if the caller reached the current range so far without having found the id and the searched id is not within the range, then we can assume that it won't be found since ranges are in the increasing order